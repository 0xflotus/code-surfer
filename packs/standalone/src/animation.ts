import { CSSProperties } from "react";
import easing, { Easing } from "./easing";
import { Tuple } from "./tuple";

const distx = 250;
const outOpacity = 0;
const outHeight = 0;

// 20% line slide to left
// 80% line change height and scroll
// 20% line slide from right
const [EXIT, SCROLL, ENTER] = [0.2, 0.8, 1];

type Animation<T> = (t: number) => T;
export type StyleAnimation = (t: number) => CSSProperties;

export function exitLine(
  fromOpacity: number,
  lineHeight: number
): StyleAnimation {
  return chain([
    [0.2, slideToLeft(fromOpacity)],
    [0.8, shrinkHeight(lineHeight)],
    [1.0, undefined]
  ]);
}

export function enterLine(toOpacity: number, lineHeight: number) {
  return chain(
    [
      [0.2, undefined],
      [0.8, growHeight(lineHeight)],
      [1.0, slideFromRight(toOpacity)]
    ],
    {
      transform: `translateX(${distx}px)`,
      height: 0,
      opacity: 0
    }
  );
}

export function focus(offOpacity: number) {
  return (t: number) => ({ opacity: tween(offOpacity, 1, t) });
}

export function unfocus(offOpacity: number) {
  return (t: number) => ({ opacity: tween(1, offOpacity, t) });
}

export function fadeOutIn(offOpacity = 0) {
  return chain([[0.5, fadeOut(offOpacity)], [1, fadeIn(offOpacity)]]);
}

export function halfFadeIn(offOpacity = 0) {
  return chain([[0.5, undefined], [1, fadeIn(offOpacity)]], { opacity: 0 });
}

export function halfFadeOut(offOpacity = 0) {
  return chain([[0.5, fadeOut(offOpacity)], [1, undefined]]);
}

export function scrollToFocus(
  t: number,
  stepPair: Tuple<{ focusCenter: number }>,
  dimensions?: { lineHeight: number }
) {
  if (!dimensions) {
    return 0;
  }

  const [prevStep, nextStep] = stepPair.spread();

  const prevCenter = prevStep
    ? prevStep.focusCenter * dimensions.lineHeight
    : 0;
  const nextCenter = nextStep
    ? nextStep.focusCenter * dimensions.lineHeight
    : 0;

  return chain(
    [
      [0.2, undefined],
      [
        0.8,
        t => ({
          scroll: tween(prevCenter, nextCenter, t, easing.easeInOutQuad)
        })
      ],
      [1, undefined]
    ],
    { scroll: prevCenter }
  )(t).scroll;
}

export function scaleToFocus(
  t: number,
  stepPair: Tuple<{
    focusCount: number;
    dimensions?: { paddingBottom: number; paddingTop: number };
  }>,
  dimensions?: {
    lineHeight: number;
    containerWidth: number;
    containerHeight: number;
    contentWidth: number;
  }
) {
  if (!dimensions) {
    return 1;
  }

  const [prev, next] = stepPair.spread();

  const prevZoom = getZoom(prev, dimensions);
  const nextZoom = getZoom(next, dimensions);

  return tween(
    prevZoom || nextZoom,
    nextZoom || prevZoom,
    t,
    easing.easeInOutQuad
  );
}

//

function slideToLeft(startOpacity: number): StyleAnimation {
  return (t: number) => ({
    opacity: tween(startOpacity, outOpacity, t),
    transform: `translateX(${tween(0, -distx, t)}px)`
  });
}

function slideFromRight(endOpacity: number): StyleAnimation {
  return (t: number) => ({
    opacity: tween(outOpacity, endOpacity, t),
    transform: `translateX(${tween(distx, 0, t)}px)`
  });
}

function shrinkHeight(lineHeight: number): StyleAnimation {
  return (t: number) => ({
    height: tween(lineHeight, outHeight, t, easing.easeInOutQuad)
  });
}

function growHeight(lineHeight: number): StyleAnimation {
  return (t: number) => ({
    height: tween(outHeight, lineHeight, t, easing.easeInOutQuad)
  });
}

export function fadeOut(offOpacity: number = 0): StyleAnimation {
  return (t: number) => ({ opacity: tween(1, offOpacity, t) });
}

export function fadeIn(offOpacity: number = 0): StyleAnimation {
  return (t: number) => ({ opacity: tween(offOpacity, 1, t) });
}

export function tween(
  from: number,
  to: number,
  t: number,
  ease: Easing = easing.linear
) {
  return from + (to - from) * ease(t);
}

export function chain<T extends object>(
  steps: [number, undefined | ((t: number) => T)][],
  start: T = {} as T
): Animation<T> {
  return (t: number) => {
    let style = start;
    let prevTop = 0;
    for (let i = 0; i < steps.length; i++) {
      const [top, fn] = steps[i];

      const stept = t > top ? 1 : (t - prevTop) / (top - prevTop);

      if (fn) {
        Object.assign(style, fn(stept));
      }

      if (t < top) {
        return style;
      }

      prevTop = top;
    }
    return style;
  };
}

function getZoom(
  step: {
    focusCount: number;
    dimensions?: { paddingBottom: number; paddingTop: number };
  },
  dimensions: {
    lineHeight: number;
    containerWidth: number;
    containerHeight: number;
    contentWidth: number;
  }
): number | null {
  if (!step) return null;

  const {
    containerHeight,
    containerWidth,
    contentWidth,
    lineHeight
  } = dimensions;

  const { paddingBottom, paddingTop } = step.dimensions;

  const contentHeight = step.focusCount * lineHeight;
  const availableHeight =
    containerHeight - Math.max(paddingBottom, paddingTop) * 2;
  const yZoom = availableHeight / contentHeight;

  // if there are lines that are too long for the container
  const xZoom = (0.9 * containerWidth) / contentWidth;

  return Math.min(yZoom, 1, xZoom);
}
