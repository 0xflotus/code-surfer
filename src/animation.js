/* eslint-disable */
import { createAnimation, Stagger, Context, run } from "./playhead/playhead";
import easing from "./playhead/easing";

const dx = 250;
const offOpacity = 0.3;

/* @jsx createAnimation */

const SlideToLeft = () => (
  <tween
    from={{ x: 0, opacity: 1 }}
    to={{ x: -dx, opacity: 0 }}
    ease={easing.easeInQuad}
  />
);

function ShrinkHeight(props, context) {
  if (!context.lineHeight) {
    return <step from={{}} to={{ height: 0 }} />;
  }
  return (
    <tween
      from={{ height: context.lineHeight }}
      to={{ height: 0 }}
      ease={easing.easeInOutQuad}
    />
  );
}

const SlideFromRight = () => (
  <tween
    from={{ x: dx, opacity: 0 }}
    to={{ x: 0, opacity: 1 }}
    ease={easing.easeOutQuad}
  />
);
function GrowHeight(props, context) {
  if (!context.lineHeight) {
    return <step from={{ height: 0 }} to={{}} />;
  }
  return (
    <tween
      from={{ height: 0 }}
      to={{ height: context.lineHeight }}
      ease={easing.easeInOutQuad}
    />
  );
}

function SwitchLines(
  { filterExit, filterEnter, prevFocus, nextFocus },
  { lines }
) {
  return (
    <parallel>
      <Stagger interval={0.2} filter={filterExit} targets={lines}>
        <chain durations={[0.35, 0.3, 0.35]}>
          <SlideToLeft />
          <ShrinkHeight />
        </chain>
      </Stagger>
      <Stagger interval={0.2} filter={filterEnter} targets={lines}>
        <chain durations={[0.35, 0.3, 0.35]}>
          <delay />
          <GrowHeight />
          <SlideFromRight />
        </chain>
      </Stagger>
      <list forEach={lines}>
        {line => (
          <tween
            from={{ opacity: prevFocus(line) ? 1 : offOpacity }}
            to={{ opacity: nextFocus(line) ? 1 : offOpacity }}
          />
        )}
      </list>
    </parallel>
  );
}

function Animation() {
  return (
    <chain durations={[0.5, 0.5]}>
      <SwitchLines
        filterExit={line => line.left && !line.middle}
        filterEnter={line => !line.left && line.middle}
        prevFocus={line => line.prevFocus}
        nextFocus={line => line.focus}
      />
      <SwitchLines
        filterExit={line => line.middle && !line.right}
        filterEnter={line => !line.middle && line.right}
        prevFocus={line => line.focus}
        nextFocus={line => line.nextFocus}
      />
    </chain>
  );
}

export function runAnimation({ lineHeight, lines, t }) {
  const animation = (
    <Context lineHeight={lineHeight} lines={lines}>
      <Animation />
    </Context>
  );
  return run(animation, t);
}

export function scrollAnimation({ t, prev, curr, next }) {
  // TODO calc params using info
  const { lineHeight, containerHeight } = curr.dimensions;

  const prevStepDims = prev ? prev.dimensions : {};
  const stepDimensions = curr.dimensions;

  const nextStepDims = next ? next.dimensions : {};

  const currentFocus = curr.focusCenter || 0;
  const prevFocus = prev ? prev.focusCenter || 0 : 0;
  const nextFocus = next ? next.focusCenter || 0 : 0;

  const currZoom = getZoom(curr, lineHeight, containerHeight, stepDimensions);
  const prevZoom =
    getZoom(prev, lineHeight, containerHeight, prevStepDims) || currZoom;
  const nextZoom =
    getZoom(next, lineHeight, containerHeight, nextStepDims) || currZoom;

  const animation = (
    <chain durations={[0.5, 0.5]}>
      <tween
        from={{
          scrollTop: prevFocus * lineHeight,
          scale: prevZoom
        }}
        to={{
          scrollTop: currentFocus * lineHeight,
          scale: currZoom
        }}
        ease={easing.easeInOutQuad}
      />
      <tween
        from={{
          scrollTop: currentFocus * lineHeight,
          scale: currZoom
        }}
        to={{
          scrollTop: nextFocus * lineHeight,
          scale: nextZoom
        }}
        ease={easing.easeInOutQuad}
      />
    </chain>
  );

  return run(animation, t);
}

function getZoom(step, lineHeight, containerHeight, stepDimensions) {
  if (!step) return null;
  const { paddingBottom, paddingTop } = stepDimensions;
  const contentHeight = step.focusCount * lineHeight;
  const availableHeight =
    containerHeight - Math.max(paddingBottom, paddingTop) * 2;
  const zoom = availableHeight / contentHeight;
  return Math.min(zoom, 1);
  // return 1;
}
