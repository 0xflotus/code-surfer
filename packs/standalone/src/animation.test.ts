import { tween, chain, exitLine, enterLine } from "./animation";
import { table } from "table";
import easing from "./easing";

test("Tween Easing", () => {
  const ts = Array(21)
    .fill(0)
    .map((_, i) => i / 20);

  const from = 10;
  const to = 20;

  const data = ts.map(t => {
    return [
      t,
      tween(from, to, t, easing.linear),
      tween(from, to, t, easing.easeInQuad),
      tween(from, to, t, easing.easeOutQuad),
      tween(from, to, t, easing.easeInOutQuad)
    ];
  });

  expect(
    toTable(
      data,
      ["t", "Linear", "In Quad", "Out Quad", "In Out Quad"],
      [, 5, 5, 5, 5]
    )
  ).toMatchSnapshot();
});

test("Chain", () => {
  const ts = Array(21)
    .fill(0)
    .map((_, i) => i / 20);

  const animation = chain([
    [0.5, (t: number) => ({ x: t, y: undefined })],
    [0.75, undefined],
    [1, (t: number) => ({ y: t, x: undefined })]
  ]);

  const data = ts.map(t => {
    const { x, y } = animation(t);
    return [t, x, y];
  });

  expect(toTable(data, ["t", "x", "y"], [, 6, 6])).toMatchSnapshot();
});

test("Line Exit", () => {
  const ts = Array(21)
    .fill(0)
    .map((_, i) => i / 20);

  const animation = exitLine(0.8, 100);
  const data = ts.map(t => {
    const { transform, height, opacity } = animation(t);
    return [t, transform, height, opacity];
  });

  expect(
    toTable(data, ["t", "transform", "height", "opacity"], [, 20, 6, 6])
  ).toMatchSnapshot();
});

test("Line Enter", () => {
  const ts = Array(21)
    .fill(0)
    .map((_, i) => i / 20);

  const animation = enterLine(0.8, 100);
  const data = ts.map(t => {
    const { transform, height, opacity } = animation(t);
    return [t, transform, height, opacity];
  });

  expect(
    toTable(data, ["t", "transform", "height", "opacity"], [, 20, 6, 6])
  ).toMatchSnapshot();
});

function toTable(data: any[][], hr: string[], truncate: number[] = []) {
  const config = {
    drawHorizontalLine: (index, size) => {
      return index === 0 || index === 1 || index === size;
    }
  };

  const newData = data.map((row, rowi) =>
    row.map((value, coli) =>
      truncate[coli]
        ? (value == null ? "" : value).toString().slice(0, truncate[coli])
        : value
    )
  );

  return "\n" + table([hr, ...newData], config);
}
