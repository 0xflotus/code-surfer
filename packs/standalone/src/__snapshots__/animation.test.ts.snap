// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chain 1`] = `
"
╔══════╤═════╤════════╗
║ t    │ x   │ y      ║
╟──────┼─────┼────────╢
║ 0    │ 0   │        ║
║ 0.05 │ 0.1 │        ║
║ 0.1  │ 0.2 │        ║
║ 0.15 │ 0.3 │        ║
║ 0.2  │ 0.4 │        ║
║ 0.25 │ 0.5 │        ║
║ 0.3  │ 0.6 │        ║
║ 0.35 │ 0.7 │        ║
║ 0.4  │ 0.8 │        ║
║ 0.45 │ 0.9 │        ║
║ 0.5  │ 1   │        ║
║ 0.55 │ 1   │        ║
║ 0.6  │ 1   │        ║
║ 0.65 │ 1   │        ║
║ 0.7  │ 1   │        ║
║ 0.75 │ 1   │ 0      ║
║ 0.8  │ 1   │ 0.2000 ║
║ 0.85 │ 1   │ 0.3999 ║
║ 0.9  │ 1   │ 0.6000 ║
║ 0.95 │ 1   │ 0.7999 ║
║ 1    │ 1   │ 1      ║
╚══════╧═════╧════════╝
"
`;

exports[`Fade In Focus 1`] = `
"
╔══════╤════════╤════════╤════════╗
║ t    │ s0     │ s1     │ s2     ║
╟──────┼────────┼────────┼────────╢
║ 0    │ 0      │ 0      │ 0      ║
║ 0.05 │ 0      │ 0      │ 0      ║
║ 0.1  │ 0      │ 0      │ 0      ║
║ 0.15 │ 0      │ 0      │ 0      ║
║ 0.2  │ 0      │ 0      │ 0      ║
║ 0.25 │ 0      │ 0      │ 0      ║
║ 0.3  │ 0      │ 0      │ 0      ║
║ 0.35 │ 0      │ 0      │ 0      ║
║ 0.4  │ 0      │ 0      │ 0      ║
║ 0.45 │ 0      │ 0      │ 0      ║
║ 0.5  │ 0      │ 0      │ 0      ║
║ 0.55 │ 0      │ 0      │ 0      ║
║ 0.6  │ 0      │ 0      │ 0      ║
║ 0.65 │ 0      │ 0      │ 0      ║
║ 0.7  │ 0      │ 0      │ 0      ║
║ 0.75 │ 0      │ 0      │ 0      ║
║ 0.8  │ 0.2857 │ 0.0714 │ 0      ║
║ 0.85 │ 0.5714 │ 0.3571 │ 0.1428 ║
║ 0.9  │ 0.8571 │ 0.6428 │ 0.4285 ║
║ 0.95 │ 1      │ 0.9285 │ 0.7142 ║
║ 1    │ 1      │ 1      │ 1      ║
╚══════╧════════╧════════╧════════╝
"
`;

exports[`Line Enter 1`] = `
"
╔══════╤══════════════════════╤════════╤═════════╗
║ t    │ transform            │ height │ opacity ║
╟──────┼──────────────────────┼────────┼─────────╢
║ 0    │ translateX(250px)    │ 0      │ 0       ║
║ 0.05 │ translateX(250px)    │ 0      │ 0       ║
║ 0.1  │ translateX(250px)    │ 0      │ 0       ║
║ 0.15 │ translateX(250px)    │ 0      │ 0       ║
║ 0.2  │ translateX(250px)    │ 0      │ 0       ║
║ 0.25 │ translateX(250px)    │ 0      │ 0       ║
║ 0.3  │ translateX(250px)    │ 0.3999 │ 0       ║
║ 0.35 │ translateX(250px)    │ 3.1999 │ 0       ║
║ 0.4  │ translateX(250px)    │ 10.800 │ 0       ║
║ 0.45 │ translateX(250px)    │ 25.600 │ 0       ║
║ 0.5  │ translateX(250px)    │ 50     │ 0       ║
║ 0.55 │ translateX(250px)    │ 74.400 │ 0       ║
║ 0.6  │ translateX(250px)    │ 89.199 │ 0       ║
║ 0.65 │ translateX(250px)    │ 96.8   │ 0       ║
║ 0.7  │ translateX(250px)    │ 99.6   │ 0       ║
║ 0.75 │ translateX(250px)    │ 100    │ 0       ║
║ 0.8  │ translateX(199.99999 │ 100    │ 0.1600  ║
║ 0.85 │ translateX(150.00000 │ 100    │ 0.3199  ║
║ 0.9  │ translateX(99.999999 │ 100    │ 0.4800  ║
║ 0.95 │ translateX(50.000000 │ 100    │ 0.6399  ║
║ 1    │ translateX(0px)      │ 100    │ 0.8     ║
╚══════╧══════════════════════╧════════╧═════════╝
"
`;

exports[`Line Exit 1`] = `
"
╔══════╤════════════════════╤════════╤═════════╗
║ t    │ transform          │ height │ opacity ║
╟──────┼────────────────────┼────────┼─────────╢
║ 0    │ translateX(0px)    │        │ 0.8     ║
║ 0.05 │ translateX(-50px)  │        │ 0.64    ║
║ 0.1  │ translateX(-100px) │        │ 0.48    ║
║ 0.15 │ translateX(-150px) │        │ 0.3200  ║
║ 0.2  │ translateX(-200px) │        │ 0.1599  ║
║ 0.25 │ translateX(-250px) │ 100    │ 0       ║
║ 0.3  │ translateX(-250px) │ 99.6   │ 0       ║
║ 0.35 │ translateX(-250px) │ 96.8   │ 0       ║
║ 0.4  │ translateX(-250px) │ 89.199 │ 0       ║
║ 0.45 │ translateX(-250px) │ 74.399 │ 0       ║
║ 0.5  │ translateX(-250px) │ 50     │ 0       ║
║ 0.55 │ translateX(-250px) │ 25.599 │ 0       ║
║ 0.6  │ translateX(-250px) │ 10.800 │ 0       ║
║ 0.65 │ translateX(-250px) │ 3.2000 │ 0       ║
║ 0.7  │ translateX(-250px) │ 0.4000 │ 0       ║
║ 0.75 │ translateX(-250px) │ 0      │ 0       ║
║ 0.8  │ translateX(-250px) │ 0      │ 0       ║
║ 0.85 │ translateX(-250px) │ 0      │ 0       ║
║ 0.9  │ translateX(-250px) │ 0      │ 0       ║
║ 0.95 │ translateX(-250px) │ 0      │ 0       ║
║ 1    │ translateX(-250px) │ 0      │ 0       ║
╚══════╧════════════════════╧════════╧═════════╝
"
`;

exports[`Stagger 1`] = `
"
╔══════╤════════╤════════╤════════╗
║ t    │ s0     │ s1     │ s2     ║
╟──────┼────────┼────────┼────────╢
║ 0    │ 0      │ 0      │ 0      ║
║ 0.05 │ 7.1428 │ 0      │ 0      ║
║ 0.1  │ 14.285 │ 0      │ 0      ║
║ 0.15 │ 21.428 │ 0      │ 0      ║
║ 0.2  │ 28.571 │ 7.1428 │ 0      ║
║ 0.25 │ 35.714 │ 14.285 │ 0      ║
║ 0.3  │ 42.857 │ 21.428 │ 0      ║
║ 0.35 │ 50     │ 28.571 │ 7.1428 ║
║ 0.4  │ 57.142 │ 35.714 │ 14.285 ║
║ 0.45 │ 64.285 │ 42.857 │ 21.428 ║
║ 0.5  │ 71.428 │ 50     │ 28.571 ║
║ 0.55 │ 78.571 │ 57.142 │ 35.714 ║
║ 0.6  │ 85.714 │ 64.285 │ 42.857 ║
║ 0.65 │ 92.857 │ 71.428 │ 50.000 ║
║ 0.7  │ 100    │ 78.571 │ 57.142 ║
║ 0.75 │ 100    │ 85.714 │ 64.285 ║
║ 0.8  │ 100    │ 92.857 │ 71.428 ║
║ 0.85 │ 100    │ 100    │ 78.571 ║
║ 0.9  │ 100    │ 100    │ 85.714 ║
║ 0.95 │ 100    │ 100    │ 92.857 ║
║ 1    │ 100    │ 100    │ 100    ║
╚══════╧════════╧════════╧════════╝
"
`;

exports[`Tween Easing 1`] = `
"
╔══════╤════════╤═════════╤══════════╤═════════════╗
║ t    │ Linear │ In Quad │ Out Quad │ In Out Quad ║
╟──────┼────────┼─────────┼──────────┼─────────────╢
║ 0    │ 10     │ 10      │ 10       │ 10          ║
║ 0.05 │ 10.5   │ 10.02   │ 10.97    │ 10.05       ║
║ 0.1  │ 11     │ 10.1    │ 11.9     │ 10.2        ║
║ 0.15 │ 11.5   │ 10.22   │ 12.77    │ 10.45       ║
║ 0.2  │ 12     │ 10.4    │ 13.60    │ 10.8        ║
║ 0.25 │ 12.5   │ 10.62   │ 14.37    │ 11.25       ║
║ 0.3  │ 13     │ 10.9    │ 15.1     │ 11.8        ║
║ 0.35 │ 13.5   │ 11.22   │ 15.77    │ 12.45       ║
║ 0.4  │ 14     │ 11.6    │ 16.40    │ 13.20       ║
║ 0.45 │ 14.5   │ 12.02   │ 16.97    │ 14.05       ║
║ 0.5  │ 15     │ 12.5    │ 17.5     │ 15          ║
║ 0.55 │ 15.5   │ 13.02   │ 17.97    │ 15.95       ║
║ 0.6  │ 16     │ 13.6    │ 18.4     │ 16.79       ║
║ 0.65 │ 16.5   │ 14.22   │ 18.77    │ 17.55       ║
║ 0.7  │ 17     │ 14.89   │ 19.1     │ 18.2        ║
║ 0.75 │ 17.5   │ 15.62   │ 19.37    │ 18.75       ║
║ 0.8  │ 18     │ 16.40   │ 19.6     │ 19.2        ║
║ 0.85 │ 18.5   │ 17.22   │ 19.77    │ 19.54       ║
║ 0.9  │ 19     │ 18.1    │ 19.9     │ 19.80       ║
║ 0.95 │ 19.5   │ 19.02   │ 19.97    │ 19.95       ║
║ 1    │ 20     │ 20      │ 20       │ 20          ║
╚══════╧════════╧═════════╧══════════╧═════════════╝
"
`;
